### PR TITLE
feat!: read/write context/permission

### LINKED ISSUE #263: Read/Write context/permission problem
This PR is realeted to https://github.com/line/cosmwasm/issues/255

dynamic link does not manage **read/write** context and permissions.

`wasmvm` (rust part) should manage **read/write** context and permission of dynamic link call stacks. 
And contracts' functions should tell `wasmvm` that it is readonly or read/write.

It is considered necessary to expose the **read/write** context and permission to the WASM binary in some way,  and for the VM to recognize it.

### PR DESCRIPTION
**Pull request by da1suk8**
# Description

This PR is related to https://github.com/line/cosmwasm/issues/263.

`#[callable_points]` macro (**NOT** `#[callable_point]`) was created to expose read-write permissions to WASM using the `_get_callable_points_properties()`.

This macro is used as follows.

```
#[callable_points]
mod callable_points {
    use super::*; 

    #[callable_point]
    fn foo(_deps: DepsMut, _env: Env, ...) {} // exposed to WASM
    
    #[callable_point]
    fn bar(_deps: Deps, _env: Env, ...) {} // exposed to WASM

    fn foobar() {} // NOT exposed to WASM
}
```
`#[callable_point]` is used as a mark rather than an attribute macro.
Only those with `#[callable_point]` above the function will be exposed to WASM.
It is also used for functions that manage read-write permissions for each function.

In the above example, the __list_callable_point() function can be used to retrieve the following map.
(`true` is read-only, `false` is read-write)
```
{
    "foo": CalleeInfo { is_read_only: false }, 
    "bar": CalleeInfo { is_read_only: true }, 
}
```
 
I also added a contract called `intermediate-number`, which has the roles of callee and caller.
Only `sub` is intentionally made a function with read-write permission.

The code expanded in macro can be viewed below.

```
$pwd
./contracts/number
$cargo expand --target wasm32-unknown-unknown
```

First, it may be necessary to add cargo expand and the target of wasm32-unknown-unknown.

- https://crates.io/crates/cargo-expand

Closes #263 

## Types of changes

- [ ] Bug fix (changes which fixes an issue)
- [x] New feature (changes which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ETC (build, ci, docs, perf, refactor, style, test)

## Checklist

- [x] I followed the [contributing guidelines](https://github.com/line/cosmwasm/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] The PR title and commits are followed [conventional commit form](https://www.conventionalcommits.org/en/v1.0.0).

### CODE REVIEW

File: packages/derive/src/callable_points.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+use crate::callable_point;
+
+pub fn make_callable_points(body: Vec<Item>) -> (Vec<TokenTree>, Vec<TokenStream>) {
+    let mut list_callable_points = Vec::new();
+    let mut maked = Vec::new();
+
+    for i in &body {
+        if let syn::Item::Fn(function) = i.clone() {
+            let (function_remove_macro, is_strip) = strip_callable_point(function.clone());
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:13]: `is_strip` -> `is_striped`

Comment (loloicci) [inline packages/derive/src/callable_points.rs:13]: can `strip_callable_point` take `&ItemFn` and return `ItemFn`?

File: packages/derive/src/callable_points.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+            if is_strip {
+                let (maked_callable_point, callee_func) =
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:16]: `maked_callable_point` -> `made_callable_point`

File: packages/derive/src/callable_points.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
+use proc_macro2::{TokenStream, TokenTree};
+use quote::quote;
+use syn::Item;
+
+use crate::callable_point;
+
+pub fn make_callable_points(body: Vec<Item>) -> (Vec<TokenTree>, Vec<TokenStream>) {
+    let mut list_callable_points = Vec::new();
+    let mut maked = Vec::new();
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:9]: `maked` -> `made`

File: packages/derive/src/callable_points.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        } else {
+            maked.extend(quote! {#i});
+        }
+    }
+
+    (maked, list_callable_points)
+}
+
+// strip #[callable_point]
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:31]: make it doc (starts with ///) and describe what is returned (for example, "tuple of 2 values. The first one is ... and the second one is ...").

File: packages/derive/src/callable_points.rs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+pub fn strip_callable_point(function: syn::ItemFn) -> (syn::ItemFn, bool) {
+    let mut res = vec![];
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:33]: It is not res(ult) itself. Rename to more describing name.

File: packages/derive/src/callable_points.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    let mut is_strip = false;
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:34]: `is_strip` -> `is_striped`

File: packages/derive/src/lib.rs [hunk 7]
```diff
@@ -183,7 +160,7 @@ pub fn callable_points(_attr: TokenStream, item: TokenStream) -> TokenStream {
                 cosmwasm_std::memory::release_buffer(vec_callee_map) as u32
             }
 
-            #(#res)*
+            #(#maked)*
```
Comment (loloicci) [inline packages/derive/src/lib.rs:163]: `maked` -> `made`

**Review by loloicci — CHANGES_REQUESTED** [review]
Write what the `callable_points` macro does in the docs. 
"Function attributed with this macro must take `deps`..." is not suitable for this macro.

https://github.com/line/cosmwasm/pull/270/commits/5d38732bd6ef4f636757dc5e3ae2e3163ee1a6dc#diff-484ed4ec90b9bea6e333d2819a67f7b7421528d49e411d19fb93f25105e070abR96

File: packages/derive/src/callable_points.rs (hunk 1 above)
Comment (da1suk8) [reply packages/derive/src/callable_points.rs:13]: why?

File: contracts/dynamic-caller-contract/src/contract.rs [hunk 8]
```diff
@@ -171,8 +171,13 @@ pub fn try_validate_interface_err(deps: Deps, _env: Env) -> Result<Response, Con
     Ok(Response::default())
 }
 
-#[callable_point]
-fn should_never_be_called(_deps: Deps, _env: Env) {}
+#[callable_points]
+mod __callable_point {
```
Comment (kokeshiM0chi) [inline contracts/dynamic-caller-contract/src/contract.rs:175]: The double underscores seem meaningless.

**Review by kokeshiM0chi — CHANGES_REQUESTED** [review]

**Comment by loloicci** [comment]
In this code, serialization of the map (name -> permission) is done in generated code. But, we can serialize it in macro (compile time) and use serialized binary in generated code.

File: packages/derive/src/lib.rs [hunk 9]
```diff
@@ -92,33 +93,93 @@ pub fn entry_point(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
-/// fn validate_address_callable_from_other_contracts(deps: Deps, _env: Env) -> Addr {
-///   // do something with deps, for example, using api.
-///   deps.api.addr_validate("dummy_human_address").unwrap()
+///     #[callable_point] // exposed to WASM
+///     fn validate_address_callable_from_other_contracts(deps: Deps, _env: Env) -> Addr {
+///         // do something with deps, for example, using api.
+///         deps.api.addr_validate("dummy_human_address").unwrap()
+///     }
+///
+///     fn foo() -> u32 { // NOT exposed to WASM
```
Comment (loloicci) [inline packages/derive/src/lib.rs:125]:
```suggestion
///     // NOT exposed to WASM
///     fn foo() -> u32 {
```

File: packages/derive/src/callable_points.rs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+/// ```
+///
+/// Output:
+///
+/// ```no_test
+/// fn foo(deps: Deps, _env: Env, ...) -> u32 {
+///     42
+/// }
+/// ```
+pub fn strip_callable_point(function: syn::ItemFn) -> (syn::ItemFn, bool) {
```
Comment (kokeshiM0chi) [inline packages/derive/src/callable_points.rs:50]: Is the file used as follows?
- `callable_point.rs` is used when handling multiple functions.
- `callable_points.rs` is used when handling one function.

If so, shouldn't this function be defined in `callable_point.rs`?

**Review by kokeshiM0chi — CHANGES_REQUESTED** [review]

File: packages/derive/src/callable_points.rs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        } else {
+            made.extend(quote! {#i});
+        }
+    }
+
+    (made, list_callable_points)
+}
+
+/// If the function has `#[callable_point]`, strip it.
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:32]: describe about 2nd value of result (bool)

File: packages/derive/src/callable_points.rs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            attrs: attrs_stripped,
+            vis: function.vis,
+            sig: function.sig,
+            block: function.block,
+        },
+        is_stripped,
+    )
+}
+
+pub fn make_callee_map_lit(list_callable_points: Vec<(String, bool)>) -> syn::LitByteStr {
```
Comment (loloicci) [inline packages/derive/src/callable_points.rs:72]: want document, especially what is the result

File: packages/derive/src/callable_point.rs [hunk 13]
```diff
@@ -86,23 +90,28 @@ pub fn make_callable_point(function: syn::ItemFn) -> TokenStream {
     } else {
         quote! { let deps = cosmwasm_std::make_dependencies() }
     };
+    (
+        quote! {
+            #[cfg(target_arch = "wasm32")]
```
Comment (kokeshiM0chi) [inline packages/derive/src/callable_point.rs:95]: About "wasm32".
Don't you think it's better to declare the character strings used for judgment together like this?

"Deps"
"Deps Mut"
"callable_point"
The same can be said for , so I'll leave it to you. Please advise if it is better to refactor in another PR.

File: contracts/intermediate-number/.gitignore [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Build results
```
Comment (kokeshiM0chi) [inline contracts/intermediate-number/.gitignore:1]: `.gitignore` also exists in the project root. The contents are the same, so isn't this file unnecessary?
https://github.com/line/cosmwasm/blob/main/.gitignore

**Review by kokeshiM0chi — CHANGES_REQUESTED** [review]

**Review by kokeshiM0chi — APPROVED** [review]
LGTM

**Comment by da1suk8** [comment]
https://github.com/line/wasmvm/pull/93#issuecomment-1475893034

The read-write permission was confusing, so the structure has been modified.
The structure is as follows. (Ex: contract of number)

```
{
    "sub": CalleeInfo { is_read_only: false }, 
    "add": CalleeInfo { is_read_only: false }, 
    "number": CalleeInfo { is_read_only: true }, 
    "mul": CalleeInfo { is_read_only: false }
}
```

**Review by loloicci — APPROVED** [review]

**Review by loloicci — CHANGES_REQUESTED** [review]
~~Sorry for canceling approval, but recheck these comments.~~

~~https://github.com/line/cosmwasm/pull/270#discussion_r1135205622~~
~~https://github.com/line/cosmwasm/pull/270#discussion_r1135233782~~
I missed https://github.com/line/cosmwasm/pull/270/commits/c337ae78827e1bff538b009bed07b091ec7e201d, sorry.
Never mind.

File: packages/derive/src/lib.rs [hunk 15]
```diff
@@ -92,33 +93,69 @@ pub fn entry_point(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
-    let maked = callable_point::make_callable_point(function);
-    res.extend(TokenStream::from(maked));
-    res
+    let (made, list_callable_points) = callable_points::make_callable_points(body);
+    let callee_map_lit = callable_points::make_callee_map_lit(list_callable_points);
+
+    let list_callable_points_ts = quote! {
+        mod #module_name {
+            #[no_mangle]
+            extern "C" fn _list_callable_points() -> u32 {
```
Comment (loloicci) [inline packages/derive/src/lib.rs:149]: How about changing the function's name to like `_get_callable_points_properties`?

**Review by loloicci — APPROVED** [review]

**Review by dudong2 — APPROVED** [review]

**Review by kokeshiM0chi — APPROVED** [review]
LGTM

**Pull request merged**